In [ ]:
print(composition_modern_minimalist)

{'Top Notes': [{'Note': 'Bergamot', 'Score': 6.0}, {'Note': 'Lemon', 'Score': 4.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Musk', 'Score': 3.0}, {'Note': 'Patchouli', 'Score': 1.0}]}


## Definir Estructuras de Datos Olfativas

Establecer las estructuras de datos (por ejemplo, diccionarios o DataFrames de pandas) para organizar las notas olfativas por familias, pirámides, características y géneros olfativos según el target.


### 💡 Design Decision

We import the pandas library, which is essential for creating and manipulating DataFrames.


In [ ]:
import pandas as pd

print("Pandas library imported successfully.")

Pandas library imported successfully.


### 💡 Design Decision

Define the olfactory categories and an initial set of example olfactory notes within a pandas DataFrame. This DataFrame will categorize notes by family, pyramid layer, characteristics, and target genre.


In [ ]:
olfactory_data = {
    'Note': [
        'Bergamot', 'Lemon', 'Pink Pepper', 'Blackcurrant', 'Apple',
        'Jasmine', 'Rose', 'Violet', 'Muguet', 'Geranium',
        'Sandalwood', 'Cedarwood', 'Vetiver', 'Amber', 'Musk',
        'Vanilla', 'Patchouli', 'Incense', 'Tonka Bean', 'Oud'
    ],
    'Family': [
        'Citrus', 'Citrus', 'Spicy', 'Fruity', 'Fruity',
        'Floral', 'Floral', 'Floral', 'Floral', 'Floral',
        'Woody', 'Woody', 'Woody', 'Oriental', 'Musky',
        'Gourmand', 'Chypre', 'Oriental', 'Gourmand', 'Oriental'
    ],
    'Pyramid_Layer': [
        'Top', 'Top', 'Top', 'Top', 'Top',
        'Middle', 'Middle', 'Middle', 'Middle', 'Middle',
        'Base', 'Base', 'Base', 'Base', 'Base',
        'Base', 'Base', 'Base', 'Base', 'Base'
    ],
    'Characteristics': [
        'Fresh', 'Zesty', 'Spicy', 'Sweet', 'Crisp',
        'Sweet', 'Romantic', 'Powdery', 'Green', 'Rosy',
        'Creamy', 'Dry', 'Earthy', 'Warm', 'Clean',
        'Sweet', 'Earthy', 'Smoky', 'Warm', 'Resinous'
    ],
    'Gender_Target': [
        'Unisex', 'Unisex', 'Unisex', 'Feminine', 'Unisex',
        'Feminine', 'Feminine', 'Feminine', 'Feminine', 'Unisex',
        'Masculine', 'Masculine', 'Masculine', 'Unisex', 'Unisex',
        'Feminine', 'Unisex', 'Masculine', 'Unisex', 'Masculine'
    ]
}

olfactory_df = pd.DataFrame(olfactory_data)
print("Olfactory data structure (DataFrame) created successfully.")
print(olfactory_df.head())


Olfactory data structure (DataFrame) created successfully.
           Note  Family Pyramid_Layer Characteristics Gender_Target
0      Bergamot  Citrus           Top           Fresh        Unisex
1         Lemon  Citrus           Top           Zesty        Unisex
2   Pink Pepper   Spicy           Top           Spicy        Unisex
3  Blackcurrant  Fruity           Top           Sweet      Feminine
4         Apple  Fruity           Top           Crisp        Unisex


## Definir Estructuras de Datos de Arquetipos

Establecer las estructuras de datos para los arquetipos de personas, incluyendo consideraciones de mercado básicas, nichos y preferencias olfativas asociadas.


### 💡 Design Decision

We define a dictionary named `archetype_data` with the specified keys and example values for various user archetypes, as per the instructions. This dictionary will hold the structured information about each archetype's market considerations, niche, and olfactory preferences.


In [ ]:
archetype_data = {
    'Modern Minimalist': {
        'Market_Considerations': ['Young Adult', 'Urban', 'Casual'],
        'Niche': 'Fresh & Clean',
        'Preferred_Families': ['Citrus', 'Green', 'Aquatic'],
        'Preferred_Characteristics': ['Fresh', 'Clean', 'Crisp'],
        'Disliked_Notes': ['Oud', 'Incense', 'Heavy Spices'],
        'Gender_Target_Preference': 'Unisex'
    },
    'Classic Romantic': {
        'Market_Considerations': ['Luxury', 'Mature Adult', 'Formal Occasions'],
        'Niche': 'Elegant & Floral',
        'Preferred_Families': ['Floral', 'Chypre', 'Oriental'],
        'Preferred_Characteristics': ['Romantic', 'Powdery', 'Sweet'],
        'Disliked_Notes': ['Vetiver', 'Strong Citrus'],
        'Gender_Target_Preference': 'Feminine'
    },
    'Adventurous Explorer': {
        'Market_Considerations': ['Middle-aged Adult', 'Niche Market', 'Unique'],
        'Niche': 'Bold & Exotic',
        'Preferred_Families': ['Woody', 'Spicy', 'Leather'],
        'Preferred_Characteristics': ['Earthy', 'Smoky', 'Resinous'],
        'Disliked_Notes': ['Muguet', 'Light Floral'],
        'Gender_Target_Preference': 'Masculine'
    },
    'Gourmand Lover': {
        'Market_Considerations': ['Young Adult', 'Trendy', 'Casual'],
        'Niche': 'Warm & Comforting',
        'Preferred_Families': ['Gourmand', 'Oriental', 'Fruity'],
        'Preferred_Characteristics': ['Sweet', 'Warm', 'Creamy'],
        'Disliked_Notes': ['Green', 'Aquatic'],
        'Gender_Target_Preference': 'Feminine'
    },
    'Elegant Professional': {
        'Market_Considerations': ['Business', 'Formal', 'Mature Adult'],
        'Niche': 'Sophisticated & Subtle',
        'Preferred_Families': ['Chypre', 'Woody', 'Musky'],
        'Preferred_Characteristics': ['Clean', 'Earthy', 'Dry'],
        'Disliked_Notes': ['Sweet', 'Fruity', 'Heavy Spices'],
        'Gender_Target_Preference': 'Unisex'
    }
}

# Convert the dictionary to a list of dictionaries for easier DataFrame creation
archetype_list = []
for name, data in archetype_data.items():
    entry = {'Archetype_Name': name}
    entry.update(data)
    archetype_list.append(entry)

archetype_df = pd.DataFrame(archetype_list)

print("Archetype data structure (DataFrame) created successfully.")
print(archetype_df.head())

Archetype data structure (DataFrame) created successfully.
         Archetype_Name                      Market_Considerations  \
0     Modern Minimalist               [Young Adult, Urban, Casual]   
1      Classic Romantic   [Luxury, Mature Adult, Formal Occasions]   
2  Adventurous Explorer  [Middle-aged Adult, Niche Market, Unique]   
3        Gourmand Lover              [Young Adult, Trendy, Casual]   
4  Elegant Professional           [Business, Formal, Mature Adult]   

                    Niche            Preferred_Families  \
0           Fresh & Clean      [Citrus, Green, Aquatic]   
1        Elegant & Floral    [Floral, Chypre, Oriental]   
2           Bold & Exotic       [Woody, Spicy, Leather]   
3       Warm & Comforting  [Gourmand, Oriental, Fruity]   
4  Sophisticated & Subtle        [Chypre, Woody, Musky]   

    Preferred_Characteristics                 Disliked_Notes  \
0       [Fresh, Clean, Crisp]   [Oud, Incense, Heavy Spices]   
1  [Romantic, Powdery, Sweet]       [

## Desarrollar Lógica de Recomendación

Diseñar la lógica para el algoritmo recomendador, que tomará un arquetipo de entrada y lo mapeará a composiciones de perfume adecuadas basándose en las estructuras de datos olfativas definidas. Esto podría incluir un sistema de puntuación o reglas de emparejamiento.


### Recommendation Logic Design

To design the recommendation logic, we will use a scoring system that evaluates each olfactory note against an input archetype's preferences and dislikes. The goal is to generate a list of suitable notes or note combinations.

#### 1. Identification using Preferred Criteria:

*   **Preferred_Families**: For each note in `olfactory_df`, we will check if its `Family` is present in the archetype's `Preferred_Families`. A match will contribute positively to the note's score.
*   **Preferred_Characteristics**: Similarly, we will check if a note's `Characteristics` match any of the archetype's `Preferred_Characteristics`. Matches will add to the score.
*   **Gender_Target_Preference**: A note's `Gender_Target` will be compared with the archetype's `Gender_Target_Preference`. Exact matches or 'Unisex' notes for a specific gender preference will receive a positive score. Mismatches will receive a zero or negative score.

#### 2. Exclusion/Penalization using Disliked Notes:

*   **Disliked_Notes**: If a note's `Note` value is found in the archetype's `Disliked_Notes`, it will receive a significant negative penalty, effectively excluding it or making its score very low.

#### 3. Scoring Mechanism:

We will implement a simple additive scoring mechanism. Each criterion met will add a predefined score to the note. The absence of a match will add zero, and a `Disliked_Note` will subtract a large value.

*   **Base Score**: Initialize each note with a score of 0.
*   **Family Match**: +3 points if `Family` is in `Preferred_Families`.
*   **Characteristic Match**: +2 points if `Characteristics` is in `Preferred_Characteristics`.
*   **Gender Match**: +1 point if `Gender_Target` matches `Gender_Target_Preference` or is 'Unisex'.
*   **Disliked Note Penalty**: -10 points if `Note` is in `Disliked_Notes`.

This will result in a ranked list of olfactory notes for the given archetype.

#### 4. Combining and Ranking Notes for Perfume Composition:

Instead of just individual notes, we aim to recommend a 'perfume composition'. This involves selecting a set of notes that complement each other across different `Pyramid_Layer`s (Top, Middle, Base).

1.  **Filter by Score**: First, filter `olfactory_df` to include only notes with a positive score based on the above mechanism.
2.  **Top Notes**: From the filtered set, select the highest-scoring Top Notes.
3.  **Middle Notes**: Select the highest-scoring Middle Notes.
4.  **Base Notes**: Select the highest-scoring Base Notes.
5.  **Composition Generation**: Combine a selection of these high-scoring notes from each layer. The number of notes from each layer can be predefined (e.g., 2 top, 3 middle, 2 base). We can then present these combinations as potential perfume profiles. Further refinement could involve ensuring variety within families or characteristics across layers to create a balanced scent. For a simpler initial approach, we can recommend a list of top-scoring individual notes, categorized by pyramid layer, and suggest that these could form a suitable composition.


### 💡 Design Decision

Now that the recommendation logic is designed, Implement a Python function to calculate scores for each olfactory note based on a given archetype's preferences and dislikes, as outlined in the previous markdown block.


In [ ]:
def get_note_scores_for_archetype(archetype_name, olfactory_df, archetype_df):
    # Get the archetype's preferences and dislikes
    archetype_preferences = archetype_df[archetype_df['Archetype_Name'] == archetype_name].iloc[0]

    preferred_families = archetype_preferences['Preferred_Families']
    preferred_characteristics = archetype_preferences['Preferred_Characteristics']
    disliked_notes = archetype_preferences['Disliked_Notes']
    gender_target_preference = archetype_preferences['Gender_Target_Preference']

    # Calculate scores for each olfactory note
    scores = []
    for index, row in olfactory_df.iterrows():
        score = 0

        # 1. Family Match
        if row['Family'] in preferred_families:
            score += 3

        # 2. Characteristic Match
        if row['Characteristics'] in preferred_characteristics:
            score += 2

        # 3. Gender Match
        if row['Gender_Target'] == gender_target_preference or row['Gender_Target'] == 'Unisex':
            score += 1
        # If the gender doesn't match and it's not unisex, penalize lightly
        elif row['Gender_Target'] != 'Unisex' and row['Gender_Target'] != gender_target_preference:
            score -= 0.5 # Small penalty for mismatching gender if not unisex

        # 4. Disliked Note Penalty
        if row['Note'] in disliked_notes:
            score -= 10

        scores.append(score)

    olfactory_df['Score'] = scores
    return olfactory_df.sort_values(by='Score', ascending=False)

print("Function 'get_note_scores_for_archetype' defined successfully.")

# Example usage for 'Modern Minimalist' archetype
scored_olfactory_notes = get_note_scores_for_archetype('Modern Minimalist', olfactory_df.copy(), archetype_df)
print("Example scored olfactory notes for 'Modern Minimalist' archetype (top 5):")
print(scored_olfactory_notes.head())

Function 'get_note_scores_for_archetype' defined successfully.
Example scored olfactory notes for 'Modern Minimalist' archetype (top 5):
         Note  Family Pyramid_Layer Characteristics Gender_Target  Score
0    Bergamot  Citrus           Top           Fresh        Unisex    6.0
1       Lemon  Citrus           Top           Zesty        Unisex    4.0
14       Musk   Musky          Base           Clean        Unisex    3.0
4       Apple  Fruity           Top           Crisp        Unisex    3.0
16  Patchouli  Chypre          Base          Earthy        Unisex    1.0


### 💡 Design Decision

Now that individual notes can be scored, Implement a function to generate perfume compositions by selecting the top-scoring notes from each pyramid layer (Top, Middle, Base) for a given archetype, as outlined in the recommendation logic design.


In [ ]:
def recommend_perfume_composition(archetype_name, olfactory_df, archetype_df, top_n_per_layer=2):
    # Get scored olfactory notes for the archetype
    scored_notes = get_note_scores_for_archetype(archetype_name, olfactory_df.copy(), archetype_df)

    # Filter for notes with a positive score
    positive_score_notes = scored_notes[scored_notes['Score'] > 0]

    if positive_score_notes.empty:
        return "No suitable notes found for this archetype with a positive score."

    recommended_composition = {
        'Top Notes': [],
        'Middle Notes': [],
        'Base Notes': []
    }

    # Select top notes for each pyramid layer
    for layer in ['Top', 'Middle', 'Base']:
        layer_notes = positive_score_notes[positive_score_notes['Pyramid_Layer'] == layer]
        # Sort by score and get top_n_per_layer
        top_notes = layer_notes.head(top_n_per_layer)
        recommended_composition[f'{layer} Notes'] = top_notes[['Note', 'Score']].to_dict(orient='records')

    return recommended_composition

print("Function 'recommend_perfume_composition' defined successfully.")

# Example usage for 'Modern Minimalist' archetype
composition_modern_minimalist = recommend_perfume_composition('Modern Minimalist', olfactory_df.copy(), archetype_df)
print("\nRecommended perfume composition for 'Modern Minimalist':")
print(composition_modern_minimalist)

# Example usage for 'Classic Romantic' archetype
composition_classic_romantic = recommend_perfume_composition('Classic Romantic', olfactory_df.copy(), archetype_df)
print("\nRecommended perfume composition for 'Classic Romantic':")
print(composition_classic_romantic)


Function 'recommend_perfume_composition' defined successfully.

Recommended perfume composition for 'Modern Minimalist':
{'Top Notes': [{'Note': 'Bergamot', 'Score': 6.0}, {'Note': 'Lemon', 'Score': 4.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Musk', 'Score': 3.0}, {'Note': 'Patchouli', 'Score': 1.0}]}

Recommended perfume composition for 'Classic Romantic':
{'Top Notes': [{'Note': 'Blackcurrant', 'Score': 3.0}, {'Note': 'Bergamot', 'Score': 1.0}], 'Middle Notes': [{'Note': 'Rose', 'Score': 6.0}, {'Note': 'Violet', 'Score': 6.0}], 'Base Notes': [{'Note': 'Amber', 'Score': 4.0}, {'Note': 'Patchouli', 'Score': 4.0}]}


## Demostrar y Refinar el Algoritmo

Probar el algoritmo con varios arquetipos de ejemplo y mostrar las recomendaciones generadas. Esto nos permitirá identificar áreas de mejora o ajustes en la lógica.


### 💡 Design Decision

I will execute the `recommend_perfume_composition` function for the 'Adventurous Explorer' and 'Gourmand Lover' archetypes and print their respective recommended compositions.


In [ ]:
composition_adventurous_explorer = recommend_perfume_composition('Adventurous Explorer', olfactory_df.copy(), archetype_df)
print("\nRecommended perfume composition for 'Adventurous Explorer':")
print(composition_adventurous_explorer)

composition_gourmand_lover = recommend_perfume_composition('Gourmand Lover', olfactory_df.copy(), archetype_df)
print("\nRecommended perfume composition for 'Gourmand Lover':")
print(composition_gourmand_lover)


Recommended perfume composition for 'Adventurous Explorer':
{'Top Notes': [{'Note': 'Pink Pepper', 'Score': 4.0}, {'Note': 'Bergamot', 'Score': 1.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Vetiver', 'Score': 6.0}, {'Note': 'Sandalwood', 'Score': 4.0}]}

Recommended perfume composition for 'Gourmand Lover':
{'Top Notes': [{'Note': 'Blackcurrant', 'Score': 6.0}, {'Note': 'Apple', 'Score': 4.0}], 'Middle Notes': [{'Note': 'Jasmine', 'Score': 3.0}, {'Note': 'Rose', 'Score': 1.0}], 'Base Notes': [{'Note': 'Amber', 'Score': 6.0}, {'Note': 'Tonka Bean', 'Score': 6.0}]}


In [ ]:
composition_elegant_professional = recommend_perfume_composition('Elegant Professional', olfactory_df.copy(), archetype_df)
print("\nRecommended perfume composition for 'Elegant Professional':")
print(composition_elegant_professional)



Recommended perfume composition for 'Elegant Professional':
{'Top Notes': [{'Note': 'Pink Pepper', 'Score': 1.0}, {'Note': 'Lemon', 'Score': 1.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Musk', 'Score': 6.0}, {'Note': 'Patchouli', 'Score': 6.0}]}


### Análisis de las Recomendaciones y Posibles Refinamientos

#### 1. Arquetipo: 'Modern Minimalist'
*   **Preferencias**: Familias `['Citrus', 'Green', 'Aquatic']`, Características `['Fresh', 'Clean', 'Crisp']`, Género `Unisex`.
*   **Recomendación**:
    *   `Top Notes`: Bergamot (Score: 6.0), Lemon (Score: 4.0)
    *   `Middle Notes`: Geranium (Score: 1.0)
    *   `Base Notes`: Musk (Score: 3.0), Patchouli (Score: 1.0)
*   **Observaciones**: Las notas superiores (Bergamota, Limón) se alinean fuertemente con las preferencias Cítricas, Frescas y Unisex. El Almizcle en las notas de base también se alinea con la característica 'Clean'. Geranio y Pachulí tienen puntuaciones más bajas, indicando una menor alineación directa, pero no son notas 'disliked'.
*   **Alineación**: Buena alineación general con las preferencias. Las notas de alta puntuación reflejan bien el perfil 'Fresh & Clean'.

#### 2. Arquetipo: 'Classic Romantic'
*   **Preferencias**: Familias `['Floral', 'Chypre', 'Oriental']`, Características `['Romantic', 'Powdery', 'Sweet']`, Género `Feminine`.
*   **Recomendación**:
    *   `Top Notes`: Blackcurrant (Score: 3.0), Bergamot (Score: 1.0)
    *   `Middle Notes`: Rose (Score: 6.0), Violet (Score: 6.0)
    *   `Base Notes`: Amber (Score: 4.0), Patchouli (Score: 4.0)
*   **Observaciones**: Las notas medias de Rosa y Violeta son un excelente ajuste, con altas puntuaciones que reflejan la preferencia Floral. El Ámbar y el Pachulí en la base también se alinean con las familias Oriental/Chypre y características dulces/cálidas. La grosella negra (Blackcurrant) aporta un toque dulce/afrutado que puede complementar, y Bergamota tiene una puntuación baja, pero no es una nota 'disliked'. Las notas fuertemente 'Floral' que no estaban directamente en las preferencias, pero sí en las características de la familia, también tuvieron una buena puntuación.
*   **Alineación**: Muy buena alineación, especialmente en las notas medias y base, que capturan la esencia 'Elegant & Floral'.

#### 3. Arquetipo: 'Adventurous Explorer'
*   **Preferencias**: Familias `['Woody', 'Spicy', 'Leather']`, Características `['Earthy', 'Smoky', 'Resinous']`, Género `Masculine`.
*   **Recomendación**:
    *   `Top Notes`: Pink Pepper (Score: 4.0), Bergamot (Score: 1.0)
    *   `Middle Notes`: Geranium (Score: 1.0)
    *   `Base Notes`: Vetiver (Score: 6.0), Sandalwood (Score: 4.0)
*   **Observaciones**: Las notas de base, Vetiver y Sándalo, se alinean muy bien con las preferencias 'Woody' y 'Earthy'. La Pimienta Rosa se alinea con la preferencia 'Spicy'. Bergamota y Geranio tienen puntuaciones bajas, lo que indica que no son un ajuste perfecto, pero se incluyen por su puntuación positiva. El algoritmo excluyó correctamente las notas 'Muguet' y 'Light Floral'.
*   **Alineación**: Buena alineación para las notas clave. El arquetipo podría beneficiarse de tener más notas 'Smoky' o 'Resinous' en el `olfactory_df` para una mayor variedad.

#### 4. Arquetipo: 'Gourmand Lover'
*   **Preferencias**: Familias `['Gourmand', 'Oriental', 'Fruity']`, Características `['Sweet', 'Warm', 'Creamy']`, Género `Feminine`.
*   **Recomendación**:
    *   `Top Notes`: Blackcurrant (Score: 6.0), Apple (Score: 4.0)
    *   `Middle Notes`: Jasmine (Score: 3.0), Rose (Score: 1.0)
    *   `Base Notes`: Amber (Score: 6.0), Tonka Bean (Score: 6.0)
*   **Observaciones**: Las recomendaciones muestran una fuerte alineación con las preferencias. Grosella Negra, Ámbar y Haba Tonka obtienen puntuaciones altas al ser 'Fruity'/'Gourmand'/'Oriental' y 'Sweet'/'Warm'. Manzana también encaja bien. Jazmín y Rosa, aunque Florales (no una familia principal), puntúan positivamente debido a sus características 'Sweet' (Jazmín) y ser 'Feminine'. El algoritmo evitó correctamente las notas 'Green' y 'Aquatic'.
*   **Alineación**: Muy buena alineación, capturando efectivamente el perfil 'Warm & Comforting' del arquetipo Gourmand.

#### Áreas de Mejora y Refinamiento:

1.  **Ampliación del `olfactory_df`**: El conjunto actual de notas es limitado. Ampliar la base de datos con más notas, especialmente aquellas que se ajustan a nichos específicos (como 'Leather' para el 'Adventurous Explorer' o más notas 'Smoky' y 'Resinous'), enriquecería las recomendaciones.
2.  **Flexibilidad de `top_n_per_layer`**: Actualmente, `top_n_per_layer` es un número fijo (2). Para arquetipos donde una capa tiene muy pocas notas con puntuación positiva, esto podría resultar en recomendaciones incompletas (ej. `Middle Notes` para 'Modern Minimalist' y 'Adventurous Explorer' solo tienen una nota). Se podría considerar una lógica para ajustar `top_n_per_layer` dinámicamente o incluir más notas con puntuaciones ligeramente más bajas si la capa está vacía o casi vacía.
3.  **Peso de las Penalizaciones**: La penalización por `Disliked_Notes` es de -10, mientras que la penalización por `Gender_Target` no coincidente es solo de -0.5. Dependiendo de la importancia que se le dé a la preferencia de género, esta penalización podría ajustarse. Si una nota no unisex y no coincidente de género debe ser fuertemente desaconsejada, la penalización podría ser mayor.
4.  **Consideración de la Armonía de Notas**: El algoritmo actual selecciona las notas con mayor puntuación de cada capa de forma independiente. Un refinamiento avanzado podría incluir una lógica para evaluar la armonía o el equilibrio entre las notas seleccionadas de diferentes capas, para asegurar una composición de perfume más coherente y agradable en lugar de solo notas individuales altamente preferidas. Por ejemplo, evitar la combinación de notas que chocan o priorizar combinaciones conocidas por ser exitosas. Esto podría requerir datos adicionales sobre combinaciones de notas exitosas o un análisis más profundo de sus compatibilidades.
5.  **Puntuación de Características**: Actualmente, se requiere una coincidencia exacta de `Characteristics`. Podría explorarse un sistema donde características similares (ej. 'Zesty' y 'Fresh' para 'Citrus') reciban una puntuación parcial, o donde un arquetipo pueda tener un conjunto más amplio de características preferidas para ser más flexible.


## Summary:

### Q&A
 The system was successfully developed by defining relevant data structures for olfactory notes and user archetypes, designing and implementing a recommendation algorithm, and demonstrating its functionality.

The developed perfume recommendation algorithm works by assigning a score to each individual olfactory note based on a specific user archetype's preferences and dislikes. It then uses these scores to create a complete perfume composition, selecting the highest-scoring notes for each pyramid layer (Top, Middle, Base).

**Capabilities of the Algorithm:**
*   **Personalized Recommendations:** It can generate tailored perfume compositions for various user archetypes based on their defined preferences (olfactory families, characteristics, gender) and explicit dislikes.
*   **Structured Output:** Recommendations are presented in a structured format, categorizing notes by their pyramid layer, which is crucial for perfume creation.
*   **Exclusion of Dislikes:** It effectively filters out notes that an archetype explicitly dislikes, ensuring that unwanted elements are not included in the recommendations.
*   **Scalability:** The framework is built on pandas DataFrames, allowing for easy expansion of both olfactory notes and user archetypes.

**Possible Extensions:**
*   **Enriched Data:** Expanding the `olfactory_df` with a broader and more diverse set of notes, including niche or less common characteristics.
*   **Dynamic Composition Size:** Implementing a more flexible approach to determine the number of notes per pyramid layer, potentially adapting based on the availability of suitable notes.
*   **Harmony and Balance:** Incorporating a mechanism to evaluate the olfactive harmony and balance between selected notes across different layers, rather than solely relying on individual note scores.
*   **Advanced Scoring:** Refining the weighting of preferences and penalties, or introducing more flexible scoring for similar characteristics (e.g., partial scores for related attributes).
*   **User Feedback Loop:** Integrating a system where user feedback on recommendations could further fine-tune archetype profiles or note scores.

### Data Analysis Key Findings
*   **Olfactory Data Structure:** A pandas DataFrame named `olfactory_df` was successfully created, categorizing 20 example olfactory notes by their `Family`, `Pyramid_Layer`, `Characteristics`, and `Gender_Target`.
*   **Archetype Data Structure:** A pandas DataFrame named `archetype_df` was established, defining four user archetypes (`Modern Minimalist`, `Classic Romantic`, `Adventurous Explorer`, `Gourmand Lover`), each with `Market_Considerations`, `Niche`, `Preferred_Families`, `Preferred_Characteristics`, `Disliked_Notes`, and `Gender_Target_Preference`.
*   **Recommendation Algorithm Logic:**
    *   An additive scoring system was implemented:
        *   +3 points for `Family` match.
        *   +2 points for `Characteristics` match.
        *   +1 point for `Gender_Target` match (or 'Unisex').
        *   -10 points penalty for `Disliked_Notes`.
        *   -0.5 points penalty for non-unisex `Gender_Target` mismatch.
    *   The algorithm generates perfume compositions by selecting the top N (defaulting to 2) positively-scoring notes for each `Pyramid_Layer` (Top, Middle, Base).
*   **Demonstrated Functionality:** The algorithm successfully generated meaningful perfume compositions for all defined archetypes, showing strong alignment with their preferences (e.g., 'Modern Minimalist' received `Bergamot` and `Lemon` as top notes, and `Musk` as a base note, reflecting 'Fresh' and 'Clean' preferences). It also correctly avoided notes explicitly listed as disliked.

### Insights or Next Steps
*   **Enrich Data for Granularity:** Expand the `olfactory_df` with a wider variety of notes, particularly those associated with specific niche characteristics (e.g., 'Smoky', 'Leather'), to provide more diverse and precise recommendations.
*   **Enhance Composition Sophistication:** Implement logic to evaluate the harmony and balance of notes across different pyramid layers, moving beyond individual note scores to create more aesthetically pleasing and coherent perfume recommendations.


## Definir Matriz de Compatibilidad de Notas

Crear una estructura de datos (por ejemplo, un diccionario o DataFrame de pandas) que defina las puntuaciones de compatibilidad entre diferentes familias olfativas o características. Esta matriz nos permitirá cuantificar qué tan bien combinan las notas de diferentes capas de la pirámide.


### 💡 Design Decision

Define a dictionary to store compatibility scores between different olfactory families, considering their common interactions in perfumery, as outlined in the instructions.


In [ ]:
olfactory_family_compatibility = {
    ('Citrus', 'Floral'): 0.8,
    ('Citrus', 'Woody'): 0.7,
    ('Citrus', 'Musky'): 0.6,
    ('Citrus', 'Spicy'): 0.5,
    ('Floral', 'Woody'): 0.8,
    ('Floral', 'Oriental'): 0.9,
    ('Floral', 'Fruity'): 0.8,
    ('Floral', 'Chypre'): 0.9,
    ('Woody', 'Oriental'): 0.7,
    ('Woody', 'Musky'): 0.8,
    ('Woody', 'Spicy'): 0.7,
    ('Woody', 'Leather'): 0.9,
    ('Oriental', 'Gourmand'): 0.9,
    ('Oriental', 'Spicy'): 0.8,
    ('Musky', 'Green'): 0.7,
    ('Musky', 'Aquatic'): 0.8,
    ('Gourmand', 'Fruity'): 0.8,
    ('Spicy', 'Fruity'): 0.6,
    ('Chypre', 'Woody'): 0.8,
    ('Green', 'Aquatic'): 0.9,
    ('Leather', 'Spicy'): 0.7,
    # Add symmetric pairs to ensure all combinations are covered if needed
    ('Floral', 'Citrus'): 0.8,
    ('Woody', 'Citrus'): 0.7,
    ('Musky', 'Citrus'): 0.6,
    ('Spicy', 'Citrus'): 0.5,
    ('Woody', 'Floral'): 0.8,
    ('Oriental', 'Floral'): 0.9,
    ('Fruity', 'Floral'): 0.8,
    ('Chypre', 'Floral'): 0.9,
    ('Oriental', 'Woody'): 0.7,
    ('Musky', 'Woody'): 0.8,
    ('Spicy', 'Woody'): 0.7,
    ('Leather', 'Woody'): 0.9,
    ('Gourmand', 'Oriental'): 0.9,
    ('Spicy', 'Oriental'): 0.8,
    ('Green', 'Musky'): 0.7,
    ('Aquatic', 'Musky'): 0.8,
    ('Fruity', 'Gourmand'): 0.8,
    ('Fruity', 'Spicy'): 0.6,
    ('Woody', 'Chypre'): 0.8,
    ('Aquatic', 'Green'): 0.9,
    ('Spicy', 'Leather'): 0.7
}

print("Dictionary 'olfactory_family_compatibility' created.")

Dictionary 'olfactory_family_compatibility' created.


### 💡 Design Decision

Now that the `olfactory_family_compatibility` dictionary is defined, I will convert it into a pandas DataFrame named `compatibility_matrix_df` with the specified columns and then display its head to verify its structure.


In [ ]:
compatibility_list = []
for (family1, family2), score in olfactory_family_compatibility.items():
    compatibility_list.append({
        'Family1': family1,
        'Family2': family2,
        'Compatibility_Score': score
    })

compatibility_matrix_df = pd.DataFrame(compatibility_list)

print("Compatibility matrix DataFrame created successfully.")
print(compatibility_matrix_df.head())

Compatibility matrix DataFrame created successfully.
  Family1 Family2  Compatibility_Score
0  Citrus  Floral                  0.8
1  Citrus   Woody                  0.7
2  Citrus   Musky                  0.6
3  Citrus   Spicy                  0.5
4  Floral   Woody                  0.8


## Incorporar Puntuación de Armonía en la Recomendación

Modificar el algoritmo de recomendación existente para integrar la matriz de compatibilidad. Esto implica desarrollar una función que calcule una 'puntuación de armonía' para una composición de perfume propuesta, basándose en la compatibilidad entre las familias olfativas de sus notas.

#### Recommendation Logic Design - Harmony Scoring

To incorporate harmony scoring, we will add a new function that evaluates the overall compatibility of a selected set of notes for a perfume composition.

#### 1. Function `calculate_harmony_score(composition, compatibility_matrix_df, olfactory_df)`:

This function will take a perfume composition (a dictionary of notes by pyramid layer), the `compatibility_matrix_df`, and the `olfactory_df` as input.

*   **Iterate through note pairs**: For each possible pair of notes within the composition, or specifically focusing on inter-layer compatibility (e.g., Top-Middle, Middle-Base, Top-Base), we will find their respective olfactory families.
*   **Retrieve compatibility score**: Look up the compatibility score between these two families in the `compatibility_matrix_df`.
    *   To handle asymmetric pairs, ensure that `compatibility_matrix_df` includes both `(Family1, Family2)` and `(Family2, Family1)` pairs, or implement a lookup that checks both.
*   **Aggregate scores**: Sum or average these compatibility scores to get an overall harmony score for the entire composition.

#### 2. Integration into `recommend_perfume_composition`:

*   After generating potential `top_n_per_layer` notes for each pyramid layer, instead of directly returning them, we will generate multiple candidate compositions.
*   For each candidate composition, we will calculate its `harmony_score` using the new function.
*   The final recommended composition will be the one that has the highest combined score (initial individual note scores + harmony score) or just the highest harmony score among the highly-rated individual notes.

#### Scoring Mechanism for Harmony:

*   For simplicity, we can calculate an average compatibility score for all unique pairs of notes across different pyramid layers in a composition.
*   For example, if a composition has notes (N1, N2, N3) from Top, Middle, Base layers respectively:
    *   Calculate compatibility between (Family(N1), Family(N2))
    *   Calculate compatibility between (Family(N2), Family(N3))
    *   Calculate compatibility between (Family(N1), Family(N3))
*   The harmony score could be the average of these compatibility scores.
*   This harmony score will then be used to refine the selection of the best perfume compositions.


### 💡 Design Decision

Now, Implement the `calculate_harmony_score` function based on the design outlined in the previous markdown block. This function will calculate an overall harmony score for a given perfume composition by averaging the compatibility scores between all unique pairs of olfactory families present in the composition.


In [ ]:
def calculate_harmony_score(composition, compatibility_matrix_df, olfactory_df):
    all_notes_in_composition = []
    for layer in composition:
        for note_info in composition[layer]:
            all_notes_in_composition.append(note_info['Note'])

    # Get the families for all notes in the composition
    families_in_composition = []
    for note in all_notes_in_composition:
        # Ensure the note exists in olfactory_df and handle potential multiple entries if needed (take the first)
        family_row = olfactory_df[olfactory_df['Note'] == note]
        if not family_row.empty:
            families_in_composition.append(family_row['Family'].iloc[0])

    # Get unique families to avoid redundant calculations for notes of the same family
    unique_families = list(set(families_in_composition))

    if len(unique_families) < 2:
        return 0.0 # Cannot calculate harmony with less than two unique families

    harmony_scores = []
    # Calculate compatibility for all unique pairs of families
    for i in range(len(unique_families)):
        for j in range(i + 1, len(unique_families)):
            family1 = unique_families[i]
            family2 = unique_families[j]

            # Look up compatibility score, checking both (family1, family2) and (family2, family1)
            score_row = compatibility_matrix_df[
                ((compatibility_matrix_df['Family1'] == family1) & (compatibility_matrix_df['Family2'] == family2)) |
                ((compatibility_matrix_df['Family1'] == family2) & (compatibility_matrix_df['Family2'] == family1))
            ]

            if not score_row.empty:
                harmony_scores.append(score_row['Compatibility_Score'].iloc[0])
            else:
                # Default low compatibility if no specific score is defined
                harmony_scores.append(0.2) # Default score for unlisted pairs

    if not harmony_scores:
        return 0.0 # No harmony scores calculated

    return sum(harmony_scores) / len(harmony_scores)

print("Function 'calculate_harmony_score' defined successfully.")

# Example usage with an existing composition (e.g., composition_modern_minimalist)
example_composition = composition_modern_minimalist
harmony = calculate_harmony_score(example_composition, compatibility_matrix_df, olfactory_df)
print(f"Harmony score for Modern Minimalist composition: {harmony}")

Function 'calculate_harmony_score' defined successfully.
Harmony score for Modern Minimalist composition: 0.4833333333333334


### Modificación de `recommend_perfume_composition` para Integrar Armonía

Para integrar la puntuación de armonía en el algoritmo de recomendación, modificaremos la función `recommend_perfume_composition` de la siguiente manera:

1.  **Generación de Candidatos**: En lugar de seleccionar directamente `top_n_per_layer` notas y formar una única composición, generaremos múltiples combinaciones de notas para cada capa (Top, Middle, Base) a partir de las notas con puntuación positiva.
    *   Para simplificar y mantener un número manejable de combinaciones, seleccionaremos las `top_k` notas con mayor puntuación para cada capa (donde `k` podría ser 2 o 3, por ejemplo). Luego, generaremos todas las combinaciones posibles tomando una nota de Top, una de Middle y una de Base.

2.  **Cálculo de Puntuación de Composición**: Para cada composición candidata generada:
    *   Calcularemos su `individual_note_score`: la suma de las puntuaciones individuales de las notas que la componen.
    *   Calcularemos su `harmony_score` utilizando la función `calculate_harmony_score` previamente definida.

3.  **Puntuación Final de Composición**: Combinaremos la puntuación individual de las notas y la puntuación de armonía para obtener una `final_composition_score`.
    *   Una forma de combinar podría ser: `final_composition_score = (peso_individual * individual_note_score) + (peso_armonia * harmony_score)`.
    *   Para esta implementación inicial, podemos simplemente sumar ambas, o dar un peso mayor a la armonía para enfatizarla.

4.  **Selección de la Mejor Composición**: La función devolverá la composición candidata que tenga la `final_composition_score` más alta.

### Consideraciones para la Generación de Candidatos:

*   **Exploración exhaustiva vs. Pragmática**: Generar *todas* las combinaciones posibles puede ser computacionalmente costoso si `top_k` es grande o si el número de capas aumenta. Para este caso, con 3 capas y `top_k=2` o `3`, el número de combinaciones será pequeño (2*2*2 = 8, o 3*3*3 = 27), lo cual es manejable.
*   **Flexibilidad en la selección**: Asegurarse de que si una capa tiene menos de `top_k` notas con puntuación positiva, se utilicen todas las disponibles. El `top_n_per_layer` original ya maneja esto hasta cierto punto.


### 💡 Design Decision

I will modify the `recommend_perfume_composition` function to incorporate the harmony scoring logic. This will involve generating candidate compositions from the top-scoring notes for each layer, calculating an overall score for each candidate (combining individual note scores and harmony scores), and then selecting the best composition.


In [ ]:
import itertools # Import itertools for generating combinations

def recommend_perfume_composition_with_harmony(archetype_name, olfactory_df, archetype_df, compatibility_matrix_df, top_k_for_combinations=3):
    # Get scored olfactory notes for the archetype
    scored_notes = get_note_scores_for_archetype(archetype_name, olfactory_df.copy(), archetype_df)

    # Filter for notes with a positive score
    positive_score_notes = scored_notes[scored_notes['Score'] > 0]

    if positive_score_notes.empty:
        return "No suitable notes found for this archetype with a positive score."

    # Get top_k notes for each pyramid layer to generate combinations
    top_notes_candidates = positive_score_notes[positive_score_notes['Pyramid_Layer'] == 'Top'].head(top_k_for_combinations)
    middle_notes_candidates = positive_score_notes[positive_score_notes['Pyramid_Layer'] == 'Middle'].head(top_k_for_combinations)
    base_notes_candidates = positive_score_notes[positive_score_notes['Pyramid_Layer'] == 'Base'].head(top_k_for_combinations)

    # If any layer is empty, return an appropriate message or handle it gracefully
    if top_notes_candidates.empty or middle_notes_candidates.empty or base_notes_candidates.empty:
        return f"Could not find enough suitable notes for all layers for archetype '{archetype_name}'."

    # Generate all possible combinations of compositions
    all_candidate_compositions = []
    for top_note_row in top_notes_candidates.itertuples():
        for middle_note_row in middle_notes_candidates.itertuples():
            for base_note_row in base_notes_candidates.itertuples():
                candidate_composition = {
                    'Top Notes': [{'Note': top_note_row.Note, 'Score': top_note_row.Score}],
                    'Middle Notes': [{'Note': middle_note_row.Note, 'Score': middle_note_row.Score}],
                    'Base Notes': [{'Note': base_note_row.Note, 'Score': base_note_row.Score}]
                }
                all_candidate_compositions.append(candidate_composition)

    if not all_candidate_compositions:
        return "No valid combinations could be generated."

    best_composition = None
    max_final_score = -1

    # Evaluate each candidate composition
    for comp in all_candidate_compositions:
        # Calculate individual note score sum
        individual_note_score_sum = 0
        for layer_notes in comp.values():
            for note_info in layer_notes:
                individual_note_score_sum += note_info['Score']

        # Calculate harmony score
        harmony_score = calculate_harmony_score(comp, compatibility_matrix_df, olfactory_df)

        # Combine scores. We can weight them if necessary. For now, simple sum + a factor for harmony
        # Giving harmony a significant weight as it's a new, important factor
        final_composition_score = individual_note_score_sum + (harmony_score * 10) # Multiply harmony by 10 to give it more impact

        if final_composition_score > max_final_score:
            max_final_score = final_composition_score
            best_composition = comp

    # Add the harmony score to the best composition for display
    if best_composition:
        best_composition['Harmony_Score'] = calculate_harmony_score(best_composition, compatibility_matrix_df, olfactory_df)
        best_composition['Final_Total_Score'] = max_final_score

    return best_composition

print("Function 'recommend_perfume_composition_with_harmony' defined successfully.")

# Example usage for 'Modern Minimalist' archetype with harmony
composition_modern_minimalist_harmony = recommend_perfume_composition_with_harmony('Modern Minimalist', olfactory_df.copy(), archetype_df, compatibility_matrix_df)
print("\nRecommended perfume composition for 'Modern Minimalist' (with harmony scoring):")
print(composition_modern_minimalist_harmony)

# Example usage for 'Classic Romantic' archetype with harmony
composition_classic_romantic_harmony = recommend_perfume_composition_with_harmony('Classic Romantic', olfactory_df.copy(), archetype_df, compatibility_matrix_df)
print("\nRecommended perfume composition for 'Classic Romantic' (with harmony scoring):")
print(composition_classic_romantic_harmony)

Function 'recommend_perfume_composition_with_harmony' defined successfully.

Recommended perfume composition for 'Modern Minimalist' (with harmony scoring):
{'Top Notes': [{'Note': 'Bergamot', 'Score': 6.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Musk', 'Score': 3.0}], 'Harmony_Score': np.float64(0.5333333333333333), 'Final_Total_Score': np.float64(15.333333333333332)}

Recommended perfume composition for 'Classic Romantic' (with harmony scoring):
{'Top Notes': [{'Note': 'Blackcurrant', 'Score': 3.0}], 'Middle Notes': [{'Note': 'Rose', 'Score': 6.0}], 'Base Notes': [{'Note': 'Amber', 'Score': 4.0}], 'Harmony_Score': np.float64(0.6333333333333334), 'Final_Total_Score': np.float64(19.333333333333336)}


### 💡 Design Decision

I will execute the `recommend_perfume_composition_with_harmony` function for the remaining archetypes ('Adventurous Explorer', 'Gourmand Lover', and 'Elegant Professional') to demonstrate the improved recommendations incorporating harmony scoring, as per the subtask requirements.


In [ ]:
composition_adventurous_explorer_harmony = recommend_perfume_composition_with_harmony('Adventurous Explorer', olfactory_df.copy(), archetype_df, compatibility_matrix_df)
print("\nRecommended perfume composition for 'Adventurous Explorer' (with harmony scoring):")
print(composition_adventurous_explorer_harmony)

composition_gourmand_lover_harmony = recommend_perfume_composition_with_harmony('Gourmand Lover', olfactory_df.copy(), archetype_df, compatibility_matrix_df)
print("\nRecommended perfume composition for 'Gourmand Lover' (with harmony scoring):")
print(composition_gourmand_lover_harmony)

composition_elegant_professional_harmony = recommend_perfume_composition_with_harmony('Elegant Professional', olfactory_df.copy(), archetype_df, compatibility_matrix_df)
print("\nRecommended perfume composition for 'Elegant Professional' (with harmony scoring):")
print(composition_elegant_professional_harmony)


Recommended perfume composition for 'Adventurous Explorer' (with harmony scoring):
{'Top Notes': [{'Note': 'Pink Pepper', 'Score': 4.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Vetiver', 'Score': 6.0}], 'Harmony_Score': np.float64(0.5666666666666667), 'Final_Total_Score': np.float64(16.666666666666664)}

Recommended perfume composition for 'Gourmand Lover' (with harmony scoring):
{'Top Notes': [{'Note': 'Blackcurrant', 'Score': 6.0}], 'Middle Notes': [{'Note': 'Jasmine', 'Score': 3.0}], 'Base Notes': [{'Note': 'Amber', 'Score': 6.0}], 'Harmony_Score': np.float64(0.6333333333333334), 'Final_Total_Score': np.float64(21.333333333333336)}

Recommended perfume composition for 'Elegant Professional' (with harmony scoring):
{'Top Notes': [{'Note': 'Lemon', 'Score': 1.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Patchouli', 'Score': 6.0}], 'Harmony_Score': np.float64(0.6333333333333333), 'Final_Total_Score': np.flo

## Summary:

### Data Analysis Key Findings

*   **Olfactory Compatibility Matrix Established**: A `compatibility_matrix_df` was successfully created, defining compatibility scores between 12 distinct olfactory families (e.g., 'Citrus', 'Floral', 'Woody', 'Musky'), including symmetric pairings, to quantify how well different notes combine.
*   **Harmony Scoring Integrated**: A `calculate_harmony_score` function was developed and integrated into the recommendation system. This function evaluates the overall compatibility of a proposed perfume composition by averaging the compatibility scores of unique olfactory family pairs within the composition. For instance, an example "Modern Minimalist" composition achieved a harmony score of 0.4833.
*   **Enhanced Recommendation Algorithm**: The `recommend_perfume_composition_with_harmony` function was implemented, which now generates multiple candidate compositions from top-scoring notes for each pyramid layer (Top, Middle, Base). Each candidate's harmony score is calculated and combined with its individual note score sum to determine a `Final_Total_Score`, with harmony scores given a significant weight (multiplied by 10) to prioritize olfactory balance.
*   **Improved Recommendations Demonstrated**: The enhanced system produced diverse and harmonized recommendations across various archetypes:
    *   **'Modern Minimalist'**: 'Bergamot', 'Geranium', 'Musk' (Harmony: 0.533, Final Score: 15.33).
    *   **'Classic Romantic'**: 'Blackcurrant', 'Rose', 'Amber' (Harmony: 0.633, Final Score: 19.33).
    *   **'Adventurous Explorer'**: 'Pink Pepper', 'Geranium', 'Vetiver' (Harmony: 0.567, Final Score: 16.67).
    *   **'Gourmand Lover'**: 'Blackcurrant', 'Jasmine', 'Amber' (Harmony: 0.633, Final Score: 21.33).
    *   **'Elegant Professional'**: 'Lemon', 'Geranium', 'Patchouli' (Harmony: 0.633, Final Score: 14.33).

### Insights or Next Steps

*   The integration of the compatibility matrix and harmony scoring significantly refines the perfume recommendation system, moving beyond individual note preferences to consider the overall balance and synergy of a composition. This leads to more sophisticated and potentially more appealing scent profiles.
*   Further enhancements could involve user feedback mechanisms to dynamically adjust compatibility scores, incorporating more complex interaction patterns (e.g., specific note-to-note compatibility rather than just family-to-family), or exploring diverse weighting strategies for the individual note scores versus the harmony score.


*   **Defining a New User Preference Structure**: Create a standardized data structure for a new user's olfactory preferences (e.g., preferred notes, families, characteristics, disliked notes, gender preference) to facilitate comparison with existing archetypes.
*   **Implementing a Fuzzy Similarity Metric**: Develop a function to calculate a 'similarity score' or 'affinity' between a new user's preferences and each existing archetype in `archetype_df`, using the overlap of preferred families, characteristics, and penalizing for disliked notes.
*   **Developing a Fuzzy Archetype Identification Function**: Create a main function that takes a new user's preferences, uses the similarity metric to score affinity with all existing archetypes, and returns the top N most similar archetypes.
*   **Modifying the Recommendation Flow**: Adjust the `recommend_perfume_composition_with_harmony` function to first identify the most suitable archetype for a given set of user preferences using the new fuzzy identification function, and then use this identified archetype to generate the harmonized perfume composition.
*   **Demonstrating Recommendation with Fuzzy Identification**: Test the system with a set of user preferences that do not exactly match any predefined archetype, showing how the algorithm identifies the closest archetype and generates a perfume recommendation, including harmony calculation.
*   **Final Task**: Summarize the implementation of fuzzy semantics for archetype identification, highlighting how it improves the adaptability of the recommendation system and discussing potential future extensions or refinements.


## Definir Estructura de Preferencias de Nuevo Usuario

Crear una estructura de datos estandarizada (similar a la de `archetype_df`) para que un nuevo usuario ingrese sus preferencias olfativas (ej. lista de notas preferidas, familias, características, notas no deseadas, género de preferencia). Esta estructura facilitará la comparación con los arquetipos existentes.


### 💡 Design Decision

Define a dictionary to store a new user's olfactory preferences, including preferred families, characteristics, disliked notes, and gender preference. These preferences will be distinct from existing archetypes to facilitate future fuzzy matching.


In [ ]:
new_user_preferences = {
    'Preferred_Families': ['Floral', 'Fruity', 'Gourmand'],
    'Preferred_Characteristics': ['Sweet', 'Romantic', 'Warm'],
    'Disliked_Notes': ['Oud', 'Vetiver', 'Incense'],
    'Gender_Target_Preference': 'Feminine'
}

print("New user preferences dictionary created successfully.")
print(new_user_preferences)

New user preferences dictionary created successfully.
{'Preferred_Families': ['Floral', 'Fruity', 'Gourmand'], 'Preferred_Characteristics': ['Sweet', 'Romantic', 'Warm'], 'Disliked_Notes': ['Oud', 'Vetiver', 'Incense'], 'Gender_Target_Preference': 'Feminine'}


## Implementar Métrica de Similitud Fuzzy

Desarrollar una función para calcular una 'puntuación de similitud' o 'afinidad' entre las preferencias de un nuevo usuario y cada arquetipo existente en `archetype_df`, utilizando la superposición de familias preferidas, características y penalizando las notas no deseadas.

#### Recommendation Logic Design - Fuzzy Similarity Metric

To implement the fuzzy similarity metric, we will create a function `calculate_fuzzy_similarity(new_user_prefs, archetype_profile)` that compares the new user's preferences with a given archetype's profile and returns a similarity score.

#### 1. Overlap Score for Preferred Families and Characteristics:

*   **Preferred_Families**: Calculate the proportion of the new user's `Preferred_Families` that are also present in the `archetype_profile['Preferred_Families']`. This will be a score between 0 and 1.
    *   `overlap_families = len(set(new_user_prefs['Preferred_Families']) & set(archetype_profile['Preferred_Families'])) / len(new_user_prefs['Preferred_Families'])`
*   **Preferred_Characteristics**: Similarly, calculate the overlap for `Preferred_Characteristics`.
    *   `overlap_characteristics = len(set(new_user_prefs['Preferred_Characteristics']) & set(archetype_profile['Preferred_Characteristics'])) / len(new_user_prefs['Preferred_Characteristics'])`

#### 2. Penalty for Disliked Notes:

*   **Disliked_Notes**: If any of the new user's `Disliked_Notes` are present in the `archetype_profile['Preferred_Families']` or `archetype_profile['Preferred_Characteristics']` (we'll need a way to map notes to families/characteristics, or directly check if archetype's preferred notes are disliked by user), a penalty will be applied.
    *   For simplicity, we can check if any `archetype_profile['Preferred_Families']` are known to be associated with user's `Disliked_Notes`' families or if archetype's `Preferred_Characteristics` align with notes user dislikes. A simpler approach for now is to check if any of the archetype's *key features* (families, characteristics) are explicitly associated with user's disliked notes.
    *   A more direct approach is to check if any notes typically associated with the archetype's `Preferred_Families` or `Preferred_Characteristics` are in the user's `Disliked_Notes` list. For the initial implementation, we will directly check if notes from `archetype_profile['Preferred_Families']` are associated with `Disliked_Notes` of the new user. Or, more simply, if any *actual notes* that would be recommended by an archetype are in the user's `Disliked_Notes`.
    *   *Refinement for Disliked Notes*: We can penalize if any of the *archetype's preferred notes* (from `archetype_profile['Preferred_Families']` or `archetype_profile['Preferred_Characteristics']`) are directly mentioned in `new_user_prefs['Disliked_Notes']` in `olfactory_df` (by looking up note families/characteristics).
    *   For a first pass, let's simplify: A penalty will be applied for each of the new user's `Disliked_Notes` that are part of the `olfactory_df` and whose 'Family' or 'Characteristics' are among the `archetype_profile`'s `Preferred_Families` or `Preferred_Characteristics` respectively. This is getting complex. Let's simplify further: a penalty if a `Disliked_Note` by the user is a prominent note in the archetype's suggested profile.
    *   **Simpler Disliked Notes Penalty**: Iterate through the `new_user_prefs['Disliked_Notes']`. For each disliked note, check its family and characteristics from `olfactory_df`. If this note's family is in the archetype's `Preferred_Families` OR its characteristics are in the archetype's `Preferred_Characteristics`, apply a penalty.

#### 3. Gender Preference Match:

*   **Gender_Target_Preference**: A score will be given if `new_user_prefs['Gender_Target_Preference']` matches `archetype_profile['Gender_Target_Preference']` or if either is 'Unisex'.

#### 4. Combining Scores:

*   The final similarity score will be a weighted sum of the overlap scores, minus any penalties.
    *   `similarity_score = (weight_families * overlap_families) + (weight_characteristics * overlap_characteristics) + (weight_gender * gender_match_score) - (weight_disliked * total_penalty_from_disliked_notes)`

#### Weights (initial proposal):
*   `weight_families = 0.4`
*   `weight_characteristics = 0.3`
*   `weight_gender = 0.2`
*   `penalty_per_disliked_overlap = 0.2`


### 💡 Design Decision

Now that the design for the fuzzy similarity metric is outlined, Implement the `calculate_fuzzy_similarity` function as described, comparing a new user's preferences against an archetype's profile based on preferred families, characteristics, disliked notes, and gender preferences.


In [ ]:
def calculate_fuzzy_similarity(new_user_prefs, archetype_profile, olfactory_df):
    similarity_score = 0

    # Weights for combining scores
    weight_families = 0.4
    weight_characteristics = 0.3
    weight_gender = 0.2
    penalty_per_disliked_overlap = 0.2

    # 1. Overlap Score for Preferred Families
    new_user_families = set(new_user_prefs['Preferred_Families'])
    archetype_families = set(archetype_profile['Preferred_Families'])
    overlap_families = len(new_user_families.intersection(archetype_families)) / len(new_user_families) if new_user_families else 0
    similarity_score += weight_families * overlap_families

    # 2. Overlap Score for Preferred Characteristics
    new_user_characteristics = set(new_user_prefs['Preferred_Characteristics'])
    archetype_characteristics = set(archetype_profile['Preferred_Characteristics'])
    overlap_characteristics = len(new_user_characteristics.intersection(archetype_characteristics)) / len(new_user_characteristics) if new_user_characteristics else 0
    similarity_score += weight_characteristics * overlap_characteristics

    # 3. Gender Preference Match
    gender_match_score = 0
    if new_user_prefs['Gender_Target_Preference'] == archetype_profile['Gender_Target_Preference']:
        gender_match_score = 1
    elif archetype_profile['Gender_Target_Preference'] == 'Unisex' or new_user_prefs['Gender_Target_Preference'] == 'Unisex':
        # If either is unisex, it's considered a partial match
        gender_match_score = 0.5
    similarity_score += weight_gender * gender_match_score

    # 4. Penalty for Disliked Notes
    disliked_penalty = 0
    for disliked_note_name in new_user_prefs['Disliked_Notes']:
        # Find the family and characteristics of the disliked note
        note_info = olfactory_df[olfactory_df['Note'] == disliked_note_name]
        if not note_info.empty:
            disliked_note_family = note_info['Family'].iloc[0]
            disliked_note_characteristic = note_info['Characteristics'].iloc[0]

            # Check if archetype's preferred families or characteristics overlap with the disliked note's attributes
            if disliked_note_family in archetype_families or disliked_note_characteristic in archetype_characteristics:
                disliked_penalty += penalty_per_disliked_overlap

    similarity_score -= disliked_penalty

    return max(0, similarity_score) # Ensure score doesn't go below zero

print("Function 'calculate_fuzzy_similarity' defined successfully.")

# Example usage for 'Modern Minimalist' archetype
modern_minimalist_profile = archetype_df[archetype_df['Archetype_Name'] == 'Modern Minimalist'].iloc[0]
fuzzy_similarity_modern_minimalist = calculate_fuzzy_similarity(new_user_preferences, modern_minimalist_profile, olfactory_df)
print(f"Fuzzy similarity for 'Modern Minimalist' archetype: {fuzzy_similarity_modern_minimalist}")

Function 'calculate_fuzzy_similarity' defined successfully.
Fuzzy similarity for 'Modern Minimalist' archetype: 0.1


## Desarrollar una Función de Identificación de Arquetipos Fuzzy

Crear una función principal que tome las preferencias de un nuevo usuario, utilice la métrica de similitud para puntuar la afinidad con todos los arquetipos existentes, y devuelva los N arquetipos más similares.

#### Recommendation Logic Design - Fuzzy Archetype Identification

We will create a function `identify_fuzzy_archetype(new_user_prefs, archetype_df, olfactory_df, top_n=1)` that takes the new user's preferences, the DataFrame of existing archetypes, the `olfactory_df` (for disliked note penalty calculation), and returns the top N most similar archetypes.

#### 1. Iterate Through Archetypes:

*   The function will loop through each archetype in the `archetype_df`.
*   For each archetype, it will extract its profile (e.g., `Preferred_Families`, `Preferred_Characteristics`, `Disliked_Notes`, `Gender_Target_Preference`).

#### 2. Calculate Similarity Score:

*   For each archetype profile, it will call the `calculate_fuzzy_similarity` function (which was just implemented) to compute a similarity score between the `new_user_prefs` and the current archetype's profile.

#### 3. Store Scores:

*   The similarity score for each archetype will be stored, along with the archetype's name, in a list of tuples or dictionaries.

#### 4. Rank and Select Top N:

*   Once all archetypes have been scored, the list will be sorted in descending order based on their similarity scores.
*   The function will then return the top `top_n` archetypes from this sorted list. If `top_n` is 1, it will return the single most similar archetype.

#### Output Structure:

The function should return a list of dictionaries, where each dictionary contains the 'Archetype_Name' and its 'Similarity_Score'.


### 💡 Design Decision

Now that the design for the fuzzy archetype identification function is outlined, Implement the `identify_fuzzy_archetype` function to score all existing archetypes against the new user's preferences and return the top N most similar archetypes.


In [ ]:
def identify_fuzzy_archetype(new_user_prefs, archetype_df, olfactory_df, top_n=1):
    archetype_similarities = []

    for index, archetype_profile in archetype_df.iterrows():
        similarity_score = calculate_fuzzy_similarity(new_user_prefs, archetype_profile, olfactory_df)
        archetype_similarities.append({
            'Archetype_Name': archetype_profile['Archetype_Name'],
            'Similarity_Score': similarity_score
        })

    # Sort archetypes by similarity score in descending order
    sorted_archetypes = sorted(archetype_similarities, key=lambda x: x['Similarity_Score'], reverse=True)

    return sorted_archetypes[:top_n]

print("Function 'identify_fuzzy_archetype' defined successfully.")

# Example usage to find the top 1 most similar archetype
top_archetype = identify_fuzzy_archetype(new_user_preferences, archetype_df.copy(), olfactory_df)
print("\nTop 1 most similar archetype for new user preferences:")
print(top_archetype)


Function 'identify_fuzzy_archetype' defined successfully.

Top 1 most similar archetype for new user preferences:
[{'Archetype_Name': 'Gourmand Lover', 'Similarity_Score': 0.2666666666666667}]


## Modificar el Flujo de Recomendación

Modificar la función `recommend_perfume_composition_with_harmony` para que primero identifique el arquetipo más adecuado para un conjunto dado de preferencias de usuario utilizando la nueva función de identificación fuzzy, y luego utilice este arquetipo identificado para generar la composición de perfume armonizada.

#### Recommendation Logic Design - Integrating Fuzzy Identification

To integrate the fuzzy archetype identification into the recommendation flow, we will create a new wrapper function, `recommend_perfume_for_new_user`, which will orchestrate the process:

1.  **Input**: This new function will take `new_user_prefs`, `archetype_df`, `olfactory_df`, `compatibility_matrix_df`, and `top_k_for_combinations` as input.
2.  **Identify Best Archetype**: It will first call `identify_fuzzy_archetype(new_user_prefs, archetype_df, olfactory_df, top_n=1)` to find the most similar archetype for the new user.
3.  **Extract Archetype Name**: From the result of `identify_fuzzy_archetype`, it will extract the `Archetype_Name` of the most similar archetype.
4.  **Generate Recommendation**: It will then call the existing `recommend_perfume_composition_with_harmony` function, passing the identified `Archetype_Name` (and other necessary dataframes) to generate the final harmonized perfume composition.

This approach keeps the core composition logic separate and reusable, while `recommend_perfume_for_new_user` handles the fuzzy mapping of a new user to an existing archetype before generating the recommendation.


### 💡 Design Decision

Now that the design for integrating fuzzy archetype identification is outlined, Implement the new wrapper function `recommend_perfume_for_new_user`. This function will first identify the most similar archetype for a new user and then use that archetype to generate a harmonized perfume composition.


In [ ]:
def recommend_perfume_for_new_user(new_user_prefs, archetype_df, olfactory_df, compatibility_matrix_df, top_k_for_combinations=3):
    # 1. Identify the most similar archetype for the new user
    identified_archetypes = identify_fuzzy_archetype(new_user_prefs, archetype_df, olfactory_df, top_n=1)

    if not identified_archetypes:
        return "No suitable archetype found for these preferences."

    # Extract the name of the most similar archetype
    best_archetype_name = identified_archetypes[0]['Archetype_Name']
    similarity_score = identified_archetypes[0]['Similarity_Score']

    print(f"Identified best archetype for new user: '{best_archetype_name}' with similarity score: {similarity_score:.2f}")

    # 2. Generate the harmonized perfume composition using the identified archetype
    # We call the existing function with the identified archetype's name
    recommendation = recommend_perfume_composition_with_harmony(
        best_archetype_name,
        olfactory_df.copy(),
        archetype_df,
        compatibility_matrix_df,
        top_k_for_combinations
    )

    return {
        'Identified_Archetype': best_archetype_name,
        'Archetype_Similarity_Score': similarity_score,
        'Recommended_Composition': recommendation
    }

print("Function 'recommend_perfume_for_new_user' defined successfully.")

Function 'recommend_perfume_for_new_user' defined successfully.


### 💡 Design Decision

I will execute the `recommend_perfume_for_new_user` function with the previously defined `new_user_preferences` to demonstrate how the system identifies the most suitable archetype and generates a harmonized perfume recommendation, as required by the subtask.


In [ ]:
new_user_recommendation = recommend_perfume_for_new_user(new_user_preferences, archetype_df.copy(), olfactory_df.copy(), compatibility_matrix_df)
print("\nRecommended perfume for new user (with fuzzy archetype identification and harmony scoring):")
print(new_user_recommendation)

Identified best archetype for new user: 'Gourmand Lover' with similarity score: 0.27

Recommended perfume for new user (with fuzzy archetype identification and harmony scoring):
{'Identified_Archetype': 'Gourmand Lover', 'Archetype_Similarity_Score': 0.2666666666666667, 'Recommended_Composition': {'Top Notes': [{'Note': 'Blackcurrant', 'Score': 6.0}], 'Middle Notes': [{'Note': 'Jasmine', 'Score': 3.0}], 'Base Notes': [{'Note': 'Amber', 'Score': 6.0}], 'Harmony_Score': np.float64(0.6333333333333334), 'Final_Total_Score': np.float64(21.333333333333336)}}


### Summary:

### Data Analysis Key Findings (After Fuzzy Archetype Identification)

*   **Fuzzy Archetype Identification Implemented**: A fuzzy semantics approach was successfully integrated to identify the most suitable user archetype based on a new user's olfactory preferences. The `calculate_fuzzy_similarity` function measures the affinity between a new user's preferences (preferred families, characteristics, disliked notes, gender preference) and existing archetypes, using weighted overlap scores and penalties.
*   **Dynamic Archetype Assignment**: The `identify_fuzzy_archetype` function now takes a new user's preferences, calculates similarity scores against all predefined archetypes, and identifies the most similar one. For example, the `new_user_preferences` (Floral, Fruity, Gourmand, Sweet, Romantic, Warm, Dislikes: Oud, Vetiver, Incense, Feminine) were identified as most similar to the 'Gourmand Lover' archetype with a similarity score of 0.27.
*   **Integrated Recommendation Flow**: The `recommend_perfume_for_new_user` function orchestrates the entire process: it first identifies the best-matching archetype for a new user and then leverages the existing `recommend_perfume_composition_with_harmony` function to generate a harmonized perfume recommendation based on that identified archetype.
*   **Adaptability Improvement**: This new fuzzy identification mechanism makes the recommendation system significantly more adaptable. Instead of requiring a user to explicitly choose an archetype, the system can now intelligently infer the closest archetype, enabling personalized recommendations even for users whose preferences don't perfectly align with any single predefined archetype.

### Capabilities of the Enhanced Algorithm:

*   **Personalized Recommendations with Dynamic Archetype Mapping**: The system can now cater to new users by automatically mapping their preferences to the most fitting archetype, providing tailored recommendations without requiring manual archetype selection.
*   **Harmonized Compositions**: The continued integration of harmony scoring ensures that recommended compositions are not only preferred by the identified archetype but also olfactively balanced and pleasing.
*   **Robustness to Diverse Preferences**: By using fuzzy matching, the system can handle a wider range of user preferences that might not perfectly match any single existing archetype.

### Possible Extensions and Refinements:

*   **Refined Fuzzy Logic**: Explore more sophisticated fuzzy logic techniques, such as fuzzy sets or different membership functions, to assign partial memberships to multiple archetypes, allowing for blended recommendations.
*   **Interactive Archetype Refinement**: Implement a feedback loop where users can confirm or adjust the identified archetype, or provide feedback on recommendations, which could then be used to fine-tune the similarity metric or archetype definitions.
*   **Weighted Preferences**: Allow users to assign weights to their preferred families or characteristics, giving them more control over what aspects of their preference are most important.
*   **Explanation of Archetype Match**: Provide a clear explanation to the user about why a particular archetype was identified as the best match, enhancing transparency and user trust.
*   **Expanding Archetype & Olfactory Data**: Continuously grow the `archetype_df` and `olfactory_df` with more diverse profiles and notes to increase the system's coverage and accuracy for a broader user base.


## Summary:

### Q&A
The developed perfume recommendation algorithm leverages a fuzzy semantics approach to enhance its adaptability. It dynamically identifies the most suitable user archetype based on new user olfactory preferences, then uses this archetype to generate harmonized perfume recommendations. This method significantly improves the system's ability to offer personalized recommendations even for users whose preferences do not perfectly align with a predefined archetype.

### Data Analysis Key Findings
*   A `new_user_preferences` dictionary was successfully created, including preferred families, characteristics, disliked notes, and gender preference, to standardize user input.
*   The `calculate_fuzzy_similarity` function was implemented to measure affinity between new user preferences and existing archetypes, using weighted overlap scores and penalties. For instance, the similarity score for the example `new_user_preferences` against the 'Modern Minimalist' archetype was 0.1.
*   The `identify_fuzzy_archetype` function successfully identified the 'Gourmand Lover' archetype as the most similar for the given `new_user_preferences` (Floral, Fruity, Gourmand, Sweet, Romantic, Warm, Dislikes: Oud, Vetiver, Incense, Feminine) with a similarity score of 0.27.
*   The `recommend_perfume_for_new_user` function integrates the fuzzy identification with the existing recommendation logic, providing a personalized, harmonized perfume composition based on the dynamically identified archetype.

### Insights or Next Steps
*   Further refine the fuzzy logic by exploring more sophisticated techniques like fuzzy sets or different membership functions to allow for partial memberships and potentially blended recommendations from multiple archetypes.
*   Implement a feedback loop that enables users to confirm or adjust the identified archetype and provide input on recommendations, which could then be used to fine-tune the similarity metric or archetype definitions.


## Demostrar Algoritmo Completo

Demonstrate the full functionality of the algorithm by defining a new user's preferences, using fuzzy identification to find the most similar archetype, and generating a harmonized perfume recommendation based on that archetype.


### 💡 Design Decision

Define a new dictionary, `another_user_preferences`, with a distinct set of olfactory preferences to represent a different user. This will allow for testing the fuzzy identification with varied input.


In [ ]:
another_user_preferences = {
    'Preferred_Families': ['Woody', 'Spicy', 'Chypre'],
    'Preferred_Characteristics': ['Earthy', 'Dry', 'Smoky'],
    'Disliked_Notes': ['Fruity', 'Gourmand', 'Sweet'],
    'Gender_Target_Preference': 'Masculine'
}

print("Another user's preferences dictionary created successfully.")
print(another_user_preferences)

Another user's preferences dictionary created successfully.
{'Preferred_Families': ['Woody', 'Spicy', 'Chypre'], 'Preferred_Characteristics': ['Earthy', 'Dry', 'Smoky'], 'Disliked_Notes': ['Fruity', 'Gourmand', 'Sweet'], 'Gender_Target_Preference': 'Masculine'}


### 💡 Design Decision

Now that `another_user_preferences` is defined, We call the `recommend_perfume_for_new_user` function with these preferences to demonstrate the fuzzy archetype identification and harmonized recommendation for a new user.


In [ ]:
another_user_recommendation = recommend_perfume_for_new_user(another_user_preferences, archetype_df.copy(), olfactory_df.copy(), compatibility_matrix_df)
print("\nRecommended perfume for another new user (with fuzzy archetype identification and harmony scoring):")
print(another_user_recommendation)

Identified best archetype for new user: 'Adventurous Explorer' with similarity score: 0.67

Recommended perfume for another new user (with fuzzy archetype identification and harmony scoring):
{'Identified_Archetype': 'Adventurous Explorer', 'Archetype_Similarity_Score': 0.6666666666666667, 'Recommended_Composition': {'Top Notes': [{'Note': 'Pink Pepper', 'Score': 4.0}], 'Middle Notes': [{'Note': 'Geranium', 'Score': 1.0}], 'Base Notes': [{'Note': 'Vetiver', 'Score': 6.0}], 'Harmony_Score': np.float64(0.5666666666666667), 'Final_Total_Score': np.float64(16.666666666666664)}}


## Resumen Final del Algoritmo

Proporcionar un resumen detallado del algoritmo de recomendación de perfumes desarrollado, destacando la integración de la semántica difusa para la identificación de arquetipos, las mejoras en la evaluación de la armonía de las notas, las capacidades actuales del sistema y las posibles extensiones a futuro.


## Summary:

### Data Analysis Key Findings

*   A new user's preferences, including `Preferred_Families` (Woody, Spicy, Chypre), `Preferred_Characteristics` (Earthy, Dry, Smoky), `Disliked_Notes` (Fruity, Gourmand, Sweet), and `Gender_Target_Preference` (Masculine), were successfully defined.
*   The fuzzy identification mechanism effectively matched these preferences to the 'Adventurous Explorer' archetype with a similarity score of approximately 0.67.
*   The system generated a harmonized perfume recommendation for the new user, featuring 'Top Notes' (Pink Pepper), 'Middle Notes' (Geranium), and 'Base Notes' (Vetiver), achieving a `Harmony_Score` of approximately 0.57 and a `Final_Total_Score` of approximately 16.67.

### Insights or Next Steps

*   The demonstration confirms the successful integration of fuzzy semantics for archetype identification and the generation of harmonized perfume recommendations based on user preferences.
*   Further enhancements could involve diversifying the predefined archetypes or incorporating user feedback loops to continuously refine the archetype matching and harmony scoring mechanisms.
